packages/peregrine/lib/List/useListState.js

```diff
@@ -0,0 +1,74 @@
+import { useReducer, useCallback, useEffect } from 'react';
+
+const updateSelection = (key, prevSelection, selectionModel) => {
+    let selection;
+    if (selectionModel === 'radio') {
+        selection = new Set().add(key);
+    }
+    if (selectionModel === 'checkbox') {
+        selection = new Set(prevSelection);
+        if (selection.has(key)) {
+            selection.delete(key);
+        } else {
+            selection.add(key);
+        }
+    }
+    return selection;
+};
+
+const reducer = (state, action) => {
```

Comment:
Could we clean this reducer up to be more like the other reducers we have? Have a look at [`useQueryResult`](https://github.com/magento-research/pwa-studio/blob/develop/packages/peregrine/lib/hooks/useQueryResult.js) for an example.

- [ ] change the second argument, `action`, to have the shape `{ payload, type }`
- [ ] wrap the reducer with `withLogger`
- [ ] memoize an `api` object with `useMemo` and return it from `useListState`
- [ ] add docblock comments

```es6
// define some initial state
const initialState = {}

// define the reducer
const reducer = (state, { payload, type }) => {
  case "": {
    // ...
  }
  default: {
    return state
  }
}

// define the hook
const useListState = () => {
  const [state, dispatch] = useReducer()

  const setFocus = useCallback(
    // ...
  )

  // memoize the api, and include dispatch
  const api = useMemo(
    () => {
      dispatch,
      setFocus
    }
  )

  return [state, api]
}
```